### PR TITLE
8312425: [vectorapi] AArch64: Optimize vector math operations with SLEEF

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review the patch? Previously it's https://github.com/openjdk/jdk/pull/18605.
This pr is based on https://github.com/openjdk/jdk/pull/20781.

Thanks!

### tests:
* test/jdk/jdk/incubator/vector/
* test/hotspot/jtreg/compiler/vectorapi/

### options:
* -XX:UseSVE=1 -XX:+EnableVectorSupport -XX:+UseVectorStubs
* -XX:UseSVE=0 -XX:+EnableVectorSupport -XX:+UseVectorStubs
* -XX:+EnableVectorSupport -XX:-UseVectorStubs

### Tests
jmh tests are test/micro/org/openjdk/bench/jdk/incubator/vector/operation/ from vectorIntrinsics branch in panama-vector. It's good to have these tests in jdk main stream, I will do it in a separate pr later. (These tests are auto-generated tests from some script&template, it's good to also have those scrip&template in jdk main stream, but those scrip&template generates more other tests than what we need here, so better to add these tests and script&template in another pr).

### Options
* +intrinsic: 'FORK=1;ITER=10;WARMUP_ITER=10;JAVA_OPTIONS=-XX:+UnlockExperimentalVMOptions -XX:+EnableVectorSupport -XX:+UseVectorStubs'
* -intrinsic: 'FORK=1;ITER=10;WARMUP_ITER=10;JAVA_OPTIONS=-XX:+UnlockExperimentalVMOptions -XX:+EnableVectorSupport -XX:-UseVectorStubs'

### Performance data
I have re-tested, there is no much difference from https://github.com/openjdk/jdk/pull/18605, so please check performance data in that pr.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (3 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 2 [Authors](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8312425](https://bugs.openjdk.org/browse/JDK-8312425): [vectorapi] AArch64: Optimize vector math operations with SLEEF (**Enhancement** - P4)

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)
 * [Fei Gao](https://openjdk.org/census#fgao) (@fg1417 - Committer)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)

### Contributors
 * Xiaohong Gong `<xgong@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21502/head:pull/21502` \
`$ git checkout pull/21502`

Update a local copy of the PR: \
`$ git checkout pull/21502` \
`$ git pull https://git.openjdk.org/jdk.git pull/21502/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21502`

View PR using the GUI difftool: \
`$ git pr show -t 21502`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21502.diff">https://git.openjdk.org/jdk/pull/21502.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21502#issuecomment-2413178791)
</details>

### CODE REVIEW

File: make/autoconf/flags-cflags.m4 [hunk 1]
```diff
@@ -912,6 +912,36 @@ AC_DEFUN([FLAGS_SETUP_CFLAGS_CPU_DEP],
         IF_FALSE: [$2FDLIBM_CFLAGS=""])
   fi
   AC_SUBST($2FDLIBM_CFLAGS)
+
+  # Check whether the compiler supports the Arm C Language Extensions (ACLE)
+  # for SVE. Set SVE_CFLAGS to -march=armv8-a+sve if it does.
+  # ACLE and this flag are required to build the aarch64 SVE related functions in
+  # libvectormath.
+  if test "x${OPENJDK_TARGET_CPU}" = "xaarch64"; then
```
Comment (magicus) [inline make/autoconf/flags-cflags.m4:920]:
```suggestion
  if test "x$OPENJDK_TARGET_CPU" = "xaarch64"; then
```

Comment (Hamlin-Li) [reply make/autoconf/flags-cflags.m4]: Thanks, Fixed

**Review by magicus — APPROVED** [review]
Build changes look fine.

/reviewers 3

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp [hunk 2]
```diff
@@ -8168,6 +8169,76 @@ class StubGenerator: public StubCodeGenerator {
...
+        }
+
+        // The native library does not support u10 level of "hypot".
+        const char* ulf = (vop == VectorSupport::VECTOR_OP_HYPOT) ? "u05" : "u10";
+
+        snprintf(ebuf, sizeof(ebuf), "%sfx_%ssve", VectorSupport::mathname[op], ulf);
+        StubRoutines::_vector_f_math[VectorSupport::VEC_SIZE_SCALABLE][op] = (address)os::dll_lookup(libsleef, ebuf);
+
+        snprintf(ebuf, sizeof(ebuf), "%sdx_%ssve", VectorSupport::mathname[op], ulf);
+        StubRoutines::_vector_d_math[VectorSupport::VEC_SIZE_SCALABLE][op] = (address)os::dll_lookup(libsleef, ebuf);
```
Comment (fg1417) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:8218]: May I ask why `aarch64` doesn't have C file including macro expansion of function names while `RISC-V` needs it, see added in https://github.com/openjdk/jdk/pull/21083/files#diff-65f5198005719e644115782e7f4dd5a17c0969b01cbb50a1224b6800bbf8f177? Thanks.

Comment (luhenry) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:8218]: Agreed, it seems to be missing `src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_sve.c` in this PR

Comment (Hamlin-Li) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:8218]: Thanks for catching!
Yes, somehow I missed this file and another one for neon in this pr when commit, will add it.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp [hunk 3]
```diff
@@ -8168,6 +8169,76 @@ class StubGenerator: public StubCodeGenerator {
...
+      }
+    }
+
+    // Math vector stubs implemented with NEON for 64/128 bits vector size.
+    for (int op = 0; op < VectorSupport::NUM_VECTOR_OP_MATH; op++) {
+      int vop = VectorSupport::VECTOR_OP_MATH_START + op;
+      if (vop == VectorSupport::VECTOR_OP_TANH) {
```
Comment (fg1417) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:8225]: Ditto

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp [hunk 4]
```diff
@@ -8168,6 +8169,76 @@ class StubGenerator: public StubCodeGenerator {
...
+    //     e.g. sinfx_u10sve is the method for computing vector float sin using SVE instructions
+    //          cosd2_u10advsimd is the method for computing 2 elements vector double cos using NEON instructions
+    //
+    log_info(library)("Loaded library %s, handle " INTPTR_FORMAT, JNI_LIB_PREFIX "sleef" JNI_LIB_SUFFIX, p2i(libsleef));
+
+    // Math vector stubs implemented with SVE for scalable vector size.
+    if (UseSVE > 0) {
+      for (int op = 0; op < VectorSupport::NUM_VECTOR_OP_MATH; op++) {
+        int vop = VectorSupport::VECTOR_OP_MATH_START + op;
+        if (vop == VectorSupport::VECTOR_OP_TANH) {
```
[the remark this answers is not in the document]
Comment (Hamlin-Li) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:8207]: Sure, will fix it. Thanks!

**Review by fg1417 — APPROVED** [review]
LGTM! Thanks for update.

**Comment by theRealAph** [comment]
Do we really have to wait for JMH tests? PRs should be reasonably reviewable, and this doesn't help.

**Review by magicus — APPROVED** [review]

**Review by theRealAph — APPROVED** [review]

**Comment by Hamlin-Li** [comment]
> Here are my results, Apple M1. Pretty similar to what we've seen, but no SVE.
> 
> Looks good.

Thank you so much for testing!

**Comment by Hamlin-Li** [comment]
Thanks all for reviewing!

/integrate

**Pull request closed without merging**